pkg/sql/sem/builtins/generator_builtins.go

```diff
@@ -1231,6 +1241,159 @@ func (g *jsonEachGenerator) Values() (tree.Datums, error) {
 	return tree.Datums{g.key, g.value}, nil
 }
 
+var jsonPopulateProps = tree.FunctionProperties{
+	Class:    tree.GeneratorClass,
+	Category: categoryGenerator,
+	// The typical way to call json_populate_record is to send NULL::atype as the
+	// first argument, so we have to accept nullable args.
+	NullableArgs: true,
+}
+
+func makeJSONPopulateImpl(gen tree.GeneratorFactory, info string) tree.Overload {
+	return tree.Overload{
+		// The json{,b}_populate_record{,set} builtins all have a 2 argument
+		// structure. The first argument is an arbitrary tuple type, which is used
+		// to set the columns of the output when the builtin is used as a FROM
+		// source, or used as-is when it's used as an ordinary projection.
+		// The second argument is a JSON object or array of objects. The builtin
+		// transforms the JSON in the second argument into the tuple in the first
+		// argument, field by field, casting fields in key "k" to the type in the
+		// tuple slot "k". Any tuple fields that were missing in the JSON will be
+		// left as they are in the input argument.
+		// The first argument can be of the form NULL::<tupletype>, in which case
+		// the default values of each field will be NULL.
+		// The second argument can also be null, in which case the first argument
+		// is returned as-is.
+		Types: tree.ArgTypes{{"base", types.Any}, {"from_json", types.Jsonb}},
+		ReturnType: func(args []tree.TypedExpr) *types.T {
+			if len(args) != 2 {
+				return tree.UnknownReturnType
+			}
+			return args[0].ResolvedType()
+		},
+		Generator: gen,
+		EvalArgs: func(ctx *tree.EvalContext, exprs tree.Exprs) (tree.Datums, error) {
+			// We need to have a hook before the function arguments are executed,
+			// because of the lack of typed null in our Datums system. The common
+			// pattern is to cast NULL to a tuple type to communicate which tuple type
+			// should be populated. This type would be erased if we simply accepted
+			// the evaluated datums as normal.
+			te := exprs[0].(tree.TypedExpr)
+			targetType := te.ResolvedType()
+			datums := make(tree.Datums, len(exprs))
+			for i := range exprs {
+				var err error
+				datums[i], err = exprs[i].(tree.TypedExpr).Eval(ctx)
+				if err != nil {
+					return nil, err
+				}
+			}
+			if datums[0] == tree.DNull {
+				// json_populate_record needs the dynamic type information in its first
+				// argument. Since we don't have typed null, and the only purpose of the
+				// first argument is to pass type information, we can synthesize an empty
+				// tuple.
```

Comment:
Synthesizing an empty tuple seems hacky. One alternative would be to change the signature of the `Fn` field of `Overload` to pass in the `FuncExpr`. Then the `FuncExpr.Exprs` could be accessed directly. Or just pass in the `Exprs` directly to `Fn`: `Fn: func(ctx *tree.EvalContext, args tree.Datums, exprs tree.Exprs) {}`.

For context, this suggestion is also motivated by me wanting to use the `func(NULL::TYPE, ....)` pattern to add a "mutation cast" builtin without having to add new keywords to the language. In this pattern, it would have to synthesize every possible type and I'm worried that it won't be trivial.

Reply:
Yeah, I guess you're right. I didn't really want to modify every `Fn` implementation if I could help it, but maybe it's worth it? The weird thing is that 99% of `Fn`s won't have to touch those `Exprs`.

I'm not sure I understand what you mean by synthesizing every possible type?

Reply:
I don't love the idea of updating the `Fn` signature for 1 or 2 usecases either.

> I'm not sure I understand what you mean by synthesizing every possible type?

In the case of `json_populate_record` you only ever need to synthesize a tuple datum with `tree.NewDTupleWithLen`. A `mutation_cast` would have to be able to synthesize a datum of every type. `mutation_cast(NULL:INT, ...)` would synthesize a `tree.DInt`, `mutation_cast(NULL::STRING, ...)` would synthesize a `tree.DString`, etc. After looking around, I did find `NewDefaultDatum`, which is similar to what we'd need for `mutation_cast`, so maybe it's not as hard as I thought.

Reply:
We could add a optional `FnExprs` attribute that takes in the exprs, that's preferred if it exists.

Reply:
I like that idea. `init` in `all_builtins.go` could assert that both `Fn` and `FnWithExprs` are not set for an overload.

Reply:
Okay, great. I will do this, it's much better than the current approach.

Reply:
![This is the way](https://c.tenor.com/2jgbEi-vCuQAAAAC/mando-way-this-is-the-way.gif)

Reply:
There's one last wrench which is that this one is a Generator function, which doesn't use Fn at all. So you'll have to redo this same thing for Fn, I think.